docs/reference/xpkg.md

```diff
@@ -0,0 +1,201 @@
+- All other content in either the `xpkg` base layer, or the full image
+  filesystem is ignored by Crossplane.
+
+> The ability to use the image's flattened filesystem is primarily for backwards
+> compatibility and is not encouraged, especially in the event that an image
+> contains more than just `xpkg` related content, due to the fact that
+> accidentally overwriting or modifying the `xpkg` layer contents in subsequent
+> layers when constructing an image could cause the package to be invalid.
+
+## package.yaml Contents
+
+Depending on the type of package, the YAML stream in the `xpkg` base layer
+`package.yaml` may contain different content. Additionally, the objects in the
+YAML stream may contain common annotations that are suitable for the given
+object type.
+
+### Configuration Package Requirements
+
+The `package.yaml` for Configuration packages must adhere to the following
+requirements:
+
+- One (1) and only one `Configuration.meta.pkg.crossplane.io` object MUST be
+  defined in the YAML stream.
+- Zero (0) or more `CompositeResourceDefinition.apiextensions.crossplane.io`
+  objects MAY be defined in the YAML stream.
+- Zero (0) or more `Composition.apiextensions.crossplane.io` objects MAY be
+  defined in the YAML stream.
+- Zero (0) other object types may be defined in the YAML stream.
+
+### Provider Package Requirements
+
+The `package.yaml` for Provider packages must adhere to the following
+requirements:
+
+- One (1) and only one `Provider.meta.pkg.crossplane.io` object MUST be defined
+  in the YAML stream.
+- Zero (0) or more `CustomResourceDefinition.apiextensions.k8s.io` objects MAY
+  be defined in the YAML stream.
+- Zero (0) other object types may be defined in the YAML stream.
+
+### Object Annotations
+
+Though not used directly by Crossplane, the following object metadata
+annotations (not to be confused with descriptor annotations in an OCI image
+manifest) are defined for `Configuration.meta.pkg.crossplane.io` and
+`Provider.meta.pkg.crossplane.io` and should be honored over any competing
+annotations by third-party consumers of Crossplane packages:
+
+- `meta.crossplane.io/maintainer`: The package's maintainers, as a short opaque
+  text string.
+- `meta.crossplane.io/source`: The URL at which the package's source can be
+  found.
+- `meta.crossplane.io/license`: The license under which the package's source is
+  released.
+- `meta.crossplane.io/description`: A one sentence description of the package.
+- `meta.crossplane.io/readme`: A longer description, documentation, etc.
+- `friendly-name.meta.crossplane.io`: A 'human-friendly' name for this package.
+- `friendly-group-name.meta.crossplane.io/<group>`: A "human-friendly" name for
+  an API group defined by this package.
+- `friendly-kind-name.meta.crossplane.io/<kind>.<group>`: A "human-friendly"
+  name for a an API kind defined by this package. Note that the kind is
+  singular, not plural (i.e. instance not instances).
```

Comment:
I considered moving this to the objects they are describing directly, but having them on the meta type (and thus the revision) make it much easier for a consumer to build a list of types and their friendly name in a given package.

Reply:
This would impose a requirement that for any add/remove/update there are two places to make changes, which is fine, but something to keep in mind because it might lead to forgetting to update this file when you add a new resource which is happening today. Though I'm not sure if we have a good way of adding an annotation to CRD through kubebuilder.

Reply:
@muvaf yeah I am not super excited about this pattern, though its worth noting that we do already employ it across providers today. These fields are optional so package functionality would not be impacted by forgetting to add. Tooling could also check for types that do not have friendly names and emit warnings if desired.

Reply:
I don't love this pattern of keeping all the friendly name annotations in the package metadata either. One thing that makes it a _little_ more tolerable in my mind is that in theory we might be able to derive friendly names in a lot of cases, so perhaps they wouldn't be needed everywhere.

Reply:
@negz would you advocate for dropping these on the CRDs / XRDs themselves and prohibiting use on the package metadata, or just supporting both? I ask because starting off with them on the package metadata does not prohibit us from eventually also supporting them on the CRDs / XRDs. However, I imagine we would only want to support `friendly-kind-name` there as we wouldn't want the same group to be defined with different friendly names on multiple CRDs / XRDs.

Reply:
FWIW, I would not be opposed to dropping annotations entirely and saying that any extra metadata is defined in 3rd-party _layer extension_ mechanism.

Reply:
> @negz would you advocate for dropping these on the CRDs / XRDs themselves and prohibiting use on the package metadata, or just supporting both?

Not necessarily. I don't particularly like either approach.

Reply:
After some offline discussion, I am going to opt to drop the friendly name annotations for the time-being as Crossplane does not use them directly for any purpose today and their inclusion on the meta package type feels somewhat clunky. If we can justify a valid usage and design in the future we can opt to add them.